src/terms/xc.jl

```diff
@@ -163,44 +163,47 @@ end
     isnothing(term.ρcore) && return nothing
 
     Vxc_real = xc_potential_real(term, basis, ψ, occupation; ρ, τ).potential
-    # TODO: move arrays to the CPU to enable forces calculations on the GPU.
-    #       Might require optimizations in the future.
     if basis.model.spin_polarization in (:none, :spinless)
-        Vxc_fourier = to_cpu(fft(basis, Vxc_real[:,:,:,1]))
+        Vxc_fourier = fft(basis, Vxc_real[:,:,:,1])
     else
-        Vxc_fourier = to_cpu(fft(basis, mean(Vxc_real, dims=4)))
+        Vxc_fourier = fft(basis, mean(Vxc_real, dims=4))
     end
 
-    model = basis.model
-    form_factors = atomic_density_form_factors(basis, CoreDensity())
+    form_factors, iG2ifnorm = atomic_density_form_factors(basis, CoreDensity())
     nlcc_groups = [(igroup, group) for (igroup, group) in enumerate(basis.model.atom_groups)
-                   if has_core_density(model.atoms[first(group)])]
+                   if has_core_density(basis.model.atoms[first(group)])]
     @assert !isnothing(nlcc_groups)
 
-    TT = promote_type(T, eltype(Vxc_real))
-    forces = [zero(Vec3{TT}) for _ = 1:length(model.positions)]
-    for (igroup, group) in nlcc_groups
-        for iatom in group
-            r = model.positions[iatom]
-            forces[iatom] = _force_xc(basis, Vxc_fourier, form_factors, igroup, r)
-        end
-    end
-    forces
+    _forces_xc(basis, Vxc_fourier, form_factors, iG2ifnorm, nlcc_groups) 
 end
 
 # Function barrier to work around various type instabilities.
-function _force_xc(basis::PlaneWaveBasis{T}, Vxc_fourier::AbstractArray{U}, form_factors,
-                   igroup, r) where {T, U}
+function _forces_xc(basis::PlaneWaveBasis{T}, Vxc_fourier::AbstractArray{U}, 
+                    form_factors, iG2ifnorm, nlcc_groups) where {T, U}
+    # Pre-allocation of large arrays for GPU Efficiency
     TT = promote_type(T, real(U))
-    f  = zero(Vec3{TT})
-    for (iG, (G, G_cart)) in enumerate(zip(to_cpu(G_vectors(basis)), to_cpu(G_vectors_cart(basis))))
-        f -= real(conj(Vxc_fourier[iG])
-                  .* form_factors[(igroup, norm(G_cart))]
-                  .* cis2pi(-dot(G, r))
-                  .* (-2T(π)) .* G .* im
-                  ./ sqrt(basis.model.unit_cell_volume))
+    Gs = G_vectors(basis)
+    work = to_device(basis.architecture, zeros(Complex{TT}, length(Gs)))
+    indices = to_device(basis.architecture, collect(1:length(Gs)))
+
+    forces = [zero(Vec3{TT}) for _ = 1:length(basis.model.positions)]
+    for (igroup, group) in nlcc_groups
+        for iatom in group
+            r = basis.model.positions[iatom]
+            ff_group = @view form_factors[:, igroup]
+            map!(work, indices) do iG
+                cis2pi(-dot(Gs[iG], r)) * conj(Vxc_fourier[iG]) * ff_group[iG2ifnorm[iG]]
+            end
+
+            forces[iatom] += map(1:3) do α
+                tmp = sum(indices) do iG
+                    -2π*im*Gs[iG][α] * work[iG]
```

Comment:
Just realised: Isn't this scalar indexing ? Why does this even work ? Is this actually compiled to a GPU kernel ... that would be pretty amazing if yes.

Reply:
This `sum()` construct does indeed run on the GPU, with a `partial_mapreduce_grid` CUDA kernel. As far as I understand, this is an alias for something like `sum(-2π*im .* Gs .* work)`, which is an array operation allowed on the GPU.
Additionally, trying to access elements of `Gs` or `work` outside of the sum leads to scalar indexing errors.